bytecode: Implement slice expressions

Rossiar opened the pull request:
Support positive and negative slices in the compiler, introduce
OpNone to allow passing empty values for the start or end of a
slice to signal the vm to default the value. The vm will return
a new user error named ErrSlice if the start index of a slice
is greater than the end index.

Co-authored-by: joshcarp <joshcarp@users.noreply.github.com>
Co-authored-by: pgmitche <pgmitche@users.noreply.github.com>

camh- commented on pkg/bytecode/code.go:
```diff
@@ -83,6 +83,14 @@ const (
+	OpSlice
+	// OpNull represents a null value. It is used as a sensible default for
+	// the start or end of an OpSlice.
+	OpNull
```
I don't think this should be called `OpNull`. Evy does not have null values so I think this can be confusing down the track. I would perhaps call it `OpNoneVal` just like the type and document it as:
> OpNoneVal represents an unspecified value used where values are optional and unspecified such as
> the start and end value of a slice index or the values in a step range.

And then update any reference to "null" values in naming to refer to "none" or similar.

camh- commented on pkg/bytecode/compiler.go:
```diff
@@ -259,6 +259,9 @@ func (c *Compiler) compileDecl(decl *parser.Decl) error {
 }
 
 func (c *Compiler) compileAssignment(stmt *parser.AssignmentStmt) error {
+	if err := c.Compile(stmt.Target); err != nil {
```
I didn't realise it was a bug to not compile the target of an assignment - I just assumed it was unimplemented because there were no more complex targets than simple variables at the time. So I don't see this as a bug fix but just extending functionality.

Since array and map values are implemented as well as the index operator, I see this change as a separate PR that allows array and map indexing on the left-hand side of an assignment. This slicing PR would follow.

The reason for this split is that 1) this functionality is not really related to slicing, and 2) it really needs more tests to test those forms of assignment. The test added in this commit does not actually test that assigning to an index works - it tests that the assignment does not affect the original value after slicing, but that test can also pass if the index assignment does not work at all. That test can be then added in the slicing PR once index assignment has been merged.

And I don't believe it does work, because the current assignment logic after this commit always creates a new global and that should not happen for array and map index assignment. It also appears that the code in the previous commit does not copy the slice so it would cause mutations in the original slice when modifying a subslice.

Rossiar replied:
> I see this change as a separate PR that allows array and map indexing on the left-hand side of an assignment

This would feel incomplete to me without being able to handle dot expressions, when I open the new PR is it reasonable to implement those in it as well?

camh- replied:
I think that should be fine.

Rossiar replied:
@camh- I went back and found that we'd need a new opcode for assigning to an index expression, so I made it part of #318 - I think that dot expressions can follow in another PR

camh- commented on pkg/bytecode/value.go:
```diff
@@ -134,6 +151,14 @@ func (a arrayVal) Index(idx value) (value, error) {
+	if err != nil {
+		return nil, err
+	}
+	return arrayVal{Elements: a.Elements[startIdx:endIdx]}, nil
```
This is not copying the slice so mutating the original slice will mutate the slice var and vice versa:

From https://docs.evy.dev/spec.html#index-and-slice : 
> The slice expression `arr[start:end]` copies a substring or subarray [...]

Evaluator implementation: https://github.com/evylang/evy/blob/main/pkg/evaluator/value.go#L232

As commented on the next commit, the test there for testing this mutation isolation does not look like it works due to index assignments not being correctly implemented.

Rossiar replied:
My bad, I guess the assignment problems obfuscated the real problem for me. I will revisit once the new PR is pushed to support assigning to index expressions.

camh- commented on pkg/bytecode/value.go:
```diff
@@ -173,3 +198,45 @@ func (m mapVal) Index(idx value) (value, error) {
+		endIdx += length
+	}
+	if startIdx > endIdx {
+		return 0, 0, fmt.Errorf("%w: %d > %d", ErrSlice, startIdx, endIdx)
```
Yeah, that's also not good. I just think if you saw an error `Invalid slice: 5 > 3` where start and end are variables and you were expecting those to hold different values (hence invalid slicing), that error message would be strange - what is `5` and what is `3`? I'm not fussed about the format of the error message WRT parens, but there should be something saying that "start is greater than end" in some way so the error is clear.

Rossiar commented on pkg/bytecode/value.go:
```diff
@@ -149,6 +166,18 @@ func (a arrayVal) Set(idx, val value) error {
+	}
+	elements := make([]value, endIdx-startIdx)
+	for i := startIdx; i < endIdx; i++ {
+		elements[i-startIdx] = a.Elements[i]
```
From my understanding we don't need to use [copyOrRef](https://github.com/evylang/evy/blob/main/pkg/evaluator/value.go#L242) here because the composite types will pass by ref and the primitives in this file are aliased onto go primitives, so will be copied by value automatically.

camh- replied:
That sounds about right. The evaluator could probably be refactored to do values the same way as you're doing here and then copyOrRef would not be needed there.

Rossiar commented on pkg/bytecode/vm.go:
```diff
@@ -160,9 +160,7 @@ func (vm *VM) Run() error {
-			if err := vm.push(m); err != nil {
-				return err
-			}
+			err = vm.push(m)
```
reducing cyclomatic complexity of `vm.Run()` - this will need a rethink and a fixup PR very soon as it is at 30/30 branches

camh- replied:
Just put in //golint:ignore whatever line. The `Run()` method is special in that it is written to be optimal and will likely need to break the rules to do so.

Rossiar commented:
@camh- I have rebased onto main and fixed array slicing so it now preserves the original value and passes the test that wrote originally (without the extra erroneous OpGetGlobals) . Please take a look!

camh- commented on pkg/bytecode/value.go:
```diff
@@ -189,3 +218,46 @@ func (m mapVal) Index(idx value) (value, error) {
+	if _, ok := start.(numVal); ok {
+		startIdx = int(start.(numVal))
+	}
+	endIdx := length
+	if _, ok := end.(numVal); ok {
+		endIdx = int(end.(numVal))
+	}
```
nit: I'd move this block of code to after the stuff with startIdx - keep all the startIdx and the endIdx stuff together.

camh- commented on pkg/bytecode/value.go:
```diff
@@ -149,6 +166,18 @@ func (a arrayVal) Set(idx, val value) error {
+		return nil, err
+	}
+	elements := make([]value, endIdx-startIdx)
+	for i := startIdx; i < endIdx; i++ {
+		elements[i-startIdx] = a.Elements[i]
+	}
```
You can use `copy` instead of the loop:
```go
copy(elements, a.Elements[startIdx:endIdx])
```
